lemminx-liberty/src/main/java/io/openliberty/tools/langserver/lemminx/LibertyDiagnosticParticipant.java

```diff
@@ -163,14 +167,31 @@ private void validateFeature(DOMDocument domDocument, List<Diagnostic> list, Set
      */
     private void checkForFeatureUniqueness(DOMDocument domDocument, List<Diagnostic> list, Set<String> includedFeatures, DOMNode featureTextNode, Set<String> versionedFeatures, Set<String> versionlessFeatures, Set<String> featuresWithoutVersions, String featureName) {
         String featureNameLower = featureName.toLowerCase();
-        String featureNameNoVersionLower=featureNameLower;
+        String featureNameNoVersionLower;
         if(featureNameLower.contains("-")){
             featureNameNoVersionLower = featureNameLower.substring(0,
                     featureNameLower.lastIndexOf("-"));
             versionedFeatures.add(featureName);
         }else{
+            featureNameNoVersionLower = featureNameLower;
             versionlessFeatures.add(featureName);
         }
+        String featureNameNoVersion = LibertyUtils.stripVersion(featureName);
+        Map<String, String> changedFeatureNameMapLower = LibertyConstants.changedFeatureNameMap.entrySet().parallelStream().collect(
+                Collectors.toMap(entry -> entry.getKey().toLowerCase(),
+                        entry -> entry.getValue().toLowerCase()));
+        Map<String, String> changedFeatureNameMapLowerReversed =
+                changedFeatureNameMapLower.entrySet()
+                        .stream()
+                        .collect(Collectors.toMap(Map.Entry::getValue, Map.Entry::getKey));
+        Set<String> featuresWithOldNames = featuresWithoutVersions.stream().map(
+                        v -> changedFeatureNameMapLower.get(v + "-"))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toSet());
+        Set<String> featuresWithChangedNames = featuresWithoutVersions.stream().map(
```

Comment:
I don't think we really need to check both directions for this. If when we find a feature with the old name, we check for existence of any features with the new name and flag the diagnostic on the old name, that should be sufficient. That way we don't have to do the reverse lookup.

Reply:
The current way of feature validation is linear and each feature will be added to included features only after each iteration. Thus we dont have a map which contains all features in the document

If we change the logic to only one way lookup,

For scenario 1, ie, new name is added first and then a old feature is coming, er the logic you suggested, this will show diagnostic 
For scenario 2, old feature is added first and then when a new name appears , per the logic you suggested, this wont show diagnostic, because we dont have a list of all features in document(we only have a list of features which is already validated)

Reply:
Got it...so it depends on which feature is found first. Does that mean we will most likely only get one message for this, but which feature gets flagged depends on their order in the list?

Reply:
btw...we could do this check after the list of features is collected at the end of the method `validateFeaturesAndPlatforms`...but it would require iterating through the includedFeatures which probably wouldn't be worth it since you can calculate the diagnostic the way you are currently.